File: cliente/cliente.py
import socket
import re

def preprocess_formula(formula):
    """
    Inserta el operador de multiplicación (*) implícito en expresiones como 1(2+3) o (2+3)(4+5).
    """
    # Agregar * entre un número y un paréntesis de apertura
    formula = re.sub(r'(\d)(\()', r'\1*\2', formula)
    # Agregar * entre un paréntesis de cierre y un número o paréntesis de apertura
    formula = re.sub(r'(\))(\d|\()', r'\1*\2', formula)
    return formula

def main():
    host = 'localhost'  # Dirección del servidor (ajustar si es necesario)
    port = 1301         # Puerto del servidor (debe coincidir con el del servidor)

    # Crear un socket para conectarse al servidor
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
        try:
            # Conectar al servidor
            client_socket.connect((host, port))
            print(f"Conectado al servidor en {host}:{port}")

            while True:
                # Solicitar al usuario una fórmula para enviar al servidor
                formula = input("Ingrese una fórmula matemática (o 'salir' para terminar): ").strip()
                if formula.lower() == 'salir':
                    print("Cerrando conexión...")
                    break

                # Preprocesar la fórmula para agregar multiplicaciones implícitas
                formula = preprocess_formula(formula)

                # Enviar la fórmula al servidor
                client_socket.sendall(formula.encode('utf-8'))

                # Recibir el resultado del servidor
                result = client_socket.recv(1024).decode('utf-8')
                print(f"Resultado recibido del servidor: {result}")

        except ConnectionRefusedError:
            print("No se pudo conectar al servidor. Asegúrese de que el servidor esté en ejecución.")
        except Exception as e:
            print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: cliente/Cliente.java
import java.io.*;
import java.net.Socket;

public class Cliente {
    private String host;
    private int port;
    private Socket socket;
    private PrintWriter out;
    private BufferedReader in;

    public Cliente(String host, int port) {
        this.host = host;
        this.port = port;
    }

    public void connect() throws IOException {
        try {
            socket = new Socket();
            socket.connect(new java.net.InetSocketAddress(host, port), 5000); // 5 seconds timeout
            out = new PrintWriter(socket.getOutputStream(), true);
            in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            System.out.println("Conectado al servidor en " + host + ":" + port);
        } catch (java.net.ConnectException e) {
            throw new IOException("No se pudo conectar al servidor. Asegúrate que el servidor esté ejecutándose en "
                    + host + ":" + port);
        } catch (java.net.SocketTimeoutException e) {
            throw new IOException("Tiempo de conexión agotado. El servidor en " + host + ":" + port + " no responde.");
        }
    }

    public String sendExpression(String expression) throws IOException {
        // Enviar la expresión al servidor (sin añadir saltos de línea)
        out.print(expression);
        out.flush(); // Forzar el envío inmediato

        // Leer la respuesta manualmente
        StringBuilder response = new StringBuilder();
        char[] buffer = new char[1024];
        int bytesRead;

        // Leer hasta que el servidor cierre la conexión o no haya más datos
        while ((bytesRead = in.read(buffer)) != -1) {
            response.append(buffer, 0, bytesRead);
        }

        return response.toString().trim();
    }

    public void disconnect() {
        try {
            if (in != null)
                in.close();
            if (out != null)
                out.close();
            if (socket != null)
                socket.close();
            System.out.println("Desconectado del servidor.");
        } catch (IOException e) {
            System.err.println("Error al desconectar " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        String host = "localhost"; // Cambia esto si el servidor está en otra máquina
        int port = 1301; // Cambia esto si el servidor escucha en otro puerto
        Cliente client = new Cliente(host, port);
        BufferedReader userInput = new BufferedReader(new InputStreamReader(System.in));

        try {
            System.out.println("Intentando conectar al servidor...");
            client.connect();

            while (true) {
                System.out.print("\nIngresa una expresión matematica (o 'EXIT' para terminar la conexión ): ");
                String expression = userInput.readLine();

                if (expression.equalsIgnoreCase("EXIT")) {
                    System.out.println("Cerrando conexión...");
                    break;
                }

                System.out.println("Enviando expresión: " + expression);
                String result = client.sendExpression(expression);
                System.out.println("Resultado: " + result);
            }
        } catch (IOException e) {
            System.err.println("Error de conexión: " + e.getMessage());
        } finally {
            client.disconnect();
        }
    }
}